pandas/core/algorithms.py

```diff
@@ -169,22 +173,24 @@ def _ensure_data(values, dtype=None):
 
     # we have failed, return object
     values = np.asarray(values, dtype=np.object)
-    return ensure_object(values), "object"
+    return ensure_object(values), np.dtype("object")
 
 
-def _reconstruct_data(values, dtype, original):
+def _reconstruct_data(
+    values: ArrayLike, dtype: DtypeObj, original: AnyArrayLike
+) -> ArrayLike:
     """
     reverse of _ensure_data
 
     Parameters
     ----------
-    values : ndarray
-    dtype : pandas_dtype
-    original : ndarray-like
+    values : np.ndarray or ExtensionArray
+    dtype : np.ndtype or ExtensionDtype
+    original : AnyArrayLike
```

Comment:
nitpick, values and dtype are 'expanded' aliases, do the same for AnyArrayLike?  

again maybe more prose, is Index only allowed with bool_dtype? what about Series?

Reply:
AFAICT it can be Index or Series regardless of dtype, whatever was passed to the top-level fucntion